Fix ArgIterator::GetArgLoc by-ref check for HFAs/HVAs

jakobbotsch opened the pull request:
This check was not taking into account that HFAs and HVAs are not passed
by ref (except for in varargs), even when they are larger than 16 bytes.
However we do not hit this bug in reflection invocation as we do not use
the size from ArgLocDesc but rather end up calling
MethodTable::GetNumInstanceFieldBytes to get the size.

cc @jkotas

jakobbotsch committed:
Fix ArgIterator::GetArgLoc by-ref check for HFAs/HVAs

This check was not taking into account that HFAs and HVAs are not passed
by ref (except for in varargs), even when they are larger than 16 bytes.
However we do not hit this bug in reflection invocation as we do not use
the size from ArgLocDesc but rather end up calling
MethodTable::GetNumInstanceFieldBytes to get the size.

jakobbotsch commented:
It looks like we could potentially hit this with COM shuffle thunks as the code to create those do access this field. Not sure how to write a test for it, however -- any pointers? cc @AaronRobinsonMSFT

AaronRobinsonMSFT commented:
> It looks like we could potentially hit this with COM shuffle thunks as the code to create those do access this field

COM doesn't support varargs. Can you point me at the location where this is used in the shuffle?

jakobbotsch commented:
> COM doesn't support varargs. Can you point me at the location where this is used in the shuffle?

What I meant in the description is that the code before was right only for HFAs/HVAs passed in varargs. It is wrong for HFAs/HVAs passed as regular arguments when there are no more float/vector registers left.

The field is used here:
https://github.com/dotnet/runtime/blob/9529c355c061abdb101adf2462902aade7474c01/src/coreclr/vm/comdelegate.cpp#L192

janvorli reviewed (approved):
LGTM, thank you!

jakobbotsch commented:
What I called COM shuffle thunks are just the normal delegate shuffle thunks now that I look closer, so it does not look related to COM interop.
With that said, it is possible to hit this bug, but it requires a lot of args. Here's a case that hits it:
```csharp
using System;

public class Program
{
    public static void Main()
    {
        Action<int, int, int, int, int, int, int, int, int, Hfa, Hfa, Hfa> act = Foo;
        act(0, 0, 0, 0, 0, 0, 0, 0, 0, new Hfa(1, 2, 3), new Hfa(4, 5, 6), new Hfa(7, 8, 9));
    }

    public static void Foo(int a, int b, int c, int d, int e, int f, int g, int h, int i, Hfa a1, Hfa a2, Hfa a3)
    {
        Console.WriteLine("{0} {1} {2}", a1, a2, a3);
    }

    public record struct Hfa(double A, double B, double C);
}
```
Diff of output with this change:
```diff
- Hfa { A = 1, B = 2, C = 3 } Hfa { A = 4, B = 5, C = 6 } Hfa { A = 7, B = 7, C = 8 }
+ Hfa { A = 1, B = 2, C = 3 } Hfa { A = 4, B = 5, C = 6 } Hfa { A = 7, B = 8, C = 9 }
```
I will investigate why the ABI stress test did not find this as it seems like it should and look into adding this as a test (either by improving ABIStress or through adding this case as a test).

jakobbotsch committed:
Add explicit test for shuffle thunks and fix hashing

The hashing was using HashCode.Add<T> with T being our custom types that
have automatically implemented GetHashCode(). Those implementations only
base the hash code on the first field in certain cases, which could hide
issues in ABIs.

jakobbotsch commented:
The reason why ABIStress did not catch this is that it is using `HashCode.Add<T>` to validate that the values make it through the calls correctly.
This ends up using the runtime-provided `GetHashCode()`, which for (at least) the HFA types only hashes the first field; with this bug, the first field is, as the only one, passed correctly.
I have pushed a fix to the test to use `HashCode.AddBytes` to hash the parameters. Now it will be interesting to see whether this hits any new failures on other platforms.

jakobbotsch committed:
Remove leftover command line argument

jakobbotsch committed:
Fix AOT as well

AaronRobinsonMSFT commented:
> What I called COM shuffle thunks are just the normal delegate shuffle thunks now that I look closer

Thanks for confirming that. I was debugging this late last night and couldn't figure out how COM interop was involved here. However, we do have some testing in this are that might be interesting to mess with. See https://github.com/dotnet/runtime/blob/9529c355c061abdb101adf2462902aade7474c01/src/tests/Interop/COM/NETClients/Primitives/NumericTests.cs#L193-L201 or search for "HFA" under `./src/tests/Interop`.

jakobbotsch commented:
Thanks @AaronRobinsonMSFT! The ABIStress covers this case now so we should have good coverage (it gets coverage of these because they are needed for pinvokes as well).
I got confused because of the class name (`COMDelegate`) and file name (`comdelegate.cpp`), but I suppose that's just a vestige of COM+.

AaronRobinsonMSFT commented:
> I got confused because of the class name (COMDelegate) and file name (comdelegate.cpp), but I suppose that's just a vestige of COM+.

Ah. Yes. That is something the Interop team has previously discussed. Perhaps we should having that conversation again.

/cc @jkoritzinsky @elinor-fung
